### PR TITLE
skip permission checks for extended attributes

### PR DESCRIPTION
**Pull request by ixhamza**
```zfs_zaccess_trivial()``` calls the ```generic_permission()``` to read xattr attributes. This causes deadlock if called from ```zpl_xattr_set_dir()``` context as xattr and the dent locks are already held in this scenario. This commit skips the permissions checks for extended attributes since the Linux VFS stack already checks it before passing us the control.

### How Has This Been Tested?

Without this patch, ```setfacl``` would be deadlocked:
```
USER_NON_ROOT=$USER
truncate -s 1G /tmp/f1
zpool create tank /tmp/f1 -f
zfs create -o xattr=dir -o acltype=posix -o mountpoint=/data tank/data
echo "TEMP" > /data/file1
chown $USER_NON_ROOT:$USER_NON_ROOT /data/file1
setfacl -m u:$USER_NON_ROOT:rwx /data/file1
```
### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by ixhamza** [comment]
@amotin, @freqlabs, @anodos325.

**Review by youzhongyang — CHANGES_REQUESTED** [review]
This is nice! I noticed the same issue one week ago but then was distracted by something else.

The fix looks good to me, I think a test case is worth to be added.

**Comment by ixhamza** [comment]
Thanks, @youzhongyang. @amotin suggested some modifications internally, I am testing those and will also add the test case.

File: module/os/linux/zfs/zpl_xattr.c [hunk 1]
```diff
@@ -433,14 +433,26 @@ zpl_xattr_get(struct inode *ip, const char *name, void *value, size_t size)
 	cred_t *cr = CRED();
 	fstrans_cookie_t cookie;
 	int error;
+	int xattr_lock = 0;
```
Comment (ryao) [inline module/os/linux/zfs/zpl_xattr.c:436]: This could be a boolean_t rather than an int.

File: module/os/linux/zfs/zfs_acl.c [hunk 2]
```diff
@@ -2465,6 +2465,8 @@ zfs_zaccess_trivial(znode_t *zp, uint32_t *working_mode, cred_t *cr,
 		return (unmapped ? SET_ERROR(EPERM) : 0);
 	}
 
+	rw_enter(&zp->z_gperm_lock, RW_READER);
```
Comment (behlendorf) [inline module/os/linux/zfs/zfs_acl.c:2468]: This lock is only ever taken as a reader and doesn't provide any protection.  Should this have been a RW_WRITER?  Was the intent here to serial calls to `generic_permission` or was it to safely set/clear `zp->z_lockedxattr `.  If the latter, couldn't we instead us the `zp->z_lock`.

Comment (ixhamza) [reply module/os/linux/zfs/zfs_acl.c:2468]: Agreed. @amotin also pointed it out and I am working on an alternative solution. The lock was supposed to protect the ```generic_permission()``` API as ```zp->z_lockedxattr``` is global to each znode. So, the write locks on every permission check do not seem to be a good idea.

File: module/os/linux/zfs/zfs_vnops_os.c [hunk 3]
```diff
@@ -555,6 +555,7 @@ zfs_create(znode_t *dzp, char *name, vattr_t *vap, int excl,
 	boolean_t	fuid_dirtied;
 	boolean_t	have_acl = B_FALSE;
 	boolean_t	waited = B_FALSE;
+	boolean_t	skipaclchk = (flag & ATTR_NOACLCHECK) ?B_TRUE :B_FALSE;
```
[the remark this answers is not in the document]
Comment (ixhamza) [reply module/os/linux/zfs/zfs_vnops_os.c:558]: I did it to keep it in one line but will change.

File: module/os/freebsd/zfs/zfs_dir.c [hunk 4]
```diff
@@ -911,7 +912,7 @@ zfs_get_xattrdir(znode_t *zp, znode_t **xzpp, cred_t *cr, int flags)
 	va.va_mode = S_IFDIR | S_ISVTX | 0777;
 	zfs_fuid_map_ids(zp, cr, &va.va_uid, &va.va_gid);
 
-	error = zfs_make_xattrdir(zp, &va, xzpp, cr);
+	error = zfs_make_xattrdir(zp, &va, xzpp, cr, B_FALSE);
```
[the remark this answers is not in the document]
Comment (ixhamza) [reply module/os/freebsd/zfs/zfs_dir.c:916]: Thanks, I will update this for consistency. The boolean is not being used for FreeBSD but was added to make the build compatible.

[the remark this answers is not in the document]
Comment (amotin) [reply module/os/freebsd/zfs/zfs_dir.c:916]: Hmm. I see Linux calls zfs_zaccess(), while FreeBSD doesn't at all, so there is nothing really to skip.  So you may ignore that comment of mine.  But now I wonder why is this there on Linux?

[the remark this answers is not in the document]
Comment (ixhamza) [reply module/os/freebsd/zfs/zfs_dir.c:916]: I checked and it was added during the initial commit while porting zfs on Linux https://github.com/openzfs/zfs/commit/34dc7c2f2#diff-5c59a2f8d51fe5cd3ff56b5d01f01a9c7cd1a28b32a61a8bbc136bbc91e18da9R821. Since Linux kernel checks the permissions within VFS stack already, I am also not very sure if it's something we really need it here or maybe it covers some edge case that I am not aware of.
@behlendorf.

File: tests/zfs-tests/tests/functional/acl/posix/posix_004_pos.ksh [hunk 5]
```diff
@@ -44,6 +44,7 @@ log_must setfacl -d -m u:$ZFS_ACL_STAFF1:rwx $TESTDIR
 log_must setfacl -b $TESTDIR
 
 log_must chown $ZFS_ACL_STAFF1 $TESTDIR
+log_must setfacl -d -m u:$ZFS_ACL_STAFF1:rwx $TESTDIR
```
Comment (freqlabs) [inline tests/zfs-tests/tests/functional/acl/posix/posix_004_pos.ksh:47]: Let's update the strategy above to explain what this is for.

**Review by freqlabs — APPROVED** [review]

File: module/os/freebsd/zfs/zfs_dir.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (ixhamza) [reply module/os/freebsd/zfs/zfs_dir.c:917]: @behlendorf Do you think if ```zfs_zaccess()``` should be kept in ```zfs_make_xattrdir()``` for Linux. FreeBSD does not have it and I am not sure if it's really required for Linux also.

[the remark this answers is not in the document]
Comment (behlendorf) [reply module/os/freebsd/zfs/zfs_dir.c:917]: @ixhamza it seems to me this shouldn't be required for Linux either, and the VFS will be performing the needed access checks.  If you can do some manual testing to make sure there's nothing obvious we're overlooking I think we can drop this.  It's been so long I don't recall why it was originally added with the original port.

[the remark this answers is not in the document]
Comment (ixhamza) [reply module/os/freebsd/zfs/zfs_dir.c:917]: @behlendorf Thank you. I did some testing by setting/retrieving extended attributes and manipulating ACLs for different users and VFS already seems to be checking all the required permissions. I have also updated the PR.

**Review by behlendorf — APPROVED** [review]

**Review by amotin — APPROVED** [review]

**Review by freqlabs — APPROVED** [review]
That worked out nicely.
